### PR TITLE
feat!: Update shouldUseFlatConfig and CLI so flat config is default

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Updates `shouldUseFlatConfig()` to return `true` unless the `ESLINT_USE_FLAT_CONFIG` environment variable is `"false"`.

Also ensures the CLI emits a warning when eslintrc is used.

Refs #13481

### CODE REVIEW

**Comment by nzakas** [comment]
Tests are failing because rules are being called with syntax that isn't supported in those older Node.js versions. Need to investigate why that's happening.

**Comment by klebba** [comment]
This PR timely! I was just about to open a related docs issue: https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file-resolution

> You can prevent this search for eslint.config.js by setting the ESLINT_USE_FLAT_CONFIG environment variable to true [...]

I think this should guide developers to set the variable to `false`

**Comment by nzakas** [comment]
@klebba documentation updates are being tracked in https://github.com/eslint/eslint/issues/13481, and that is one of the documentation changes we'll be making.

File: tests/lib/cli.js [hunk 1]
```diff
@@ -371,7 +370,7 @@ describe("cli", () => {
                 it(`should execute without any errors with configType:${configType}`, async () => {
                     const formatterPath = getFixturePath("formatters", "async.js");
                     const filePath = getFixturePath("passing.js");
-                    const exit = await cli.execute(`-f ${formatterPath} ${filePath}`);
+                    const exit = await cli.execute(`-f ${formatterPath} ${filePath} --no-config-lookup`);
```
Comment (mdjermanovic) [inline tests/lib/cli.js:373]:
```suggestion
                    const exit = await cli.execute(`-f ${formatterPath} ${filePath} ${flag}`, null, useFlatConfig);
```

To test both modes.

File: tests/lib/cli.js [hunk 2]
```diff
@@ -1484,17 +1483,40 @@ describe("cli", () => {
...
+
+            afterEach(() => {
+                processStub.restore();
+            });
+
+            it("should emit deprecation warning", async () => {
+                const configPath = getFixturePath(".eslintrc");
+                const filePath = getFixturePath("passing.js");
+
+                await cli.execute(`--config ${configPath} ${filePath}`, null, false);
```
[the remark this answers is not in the document]
Comment (nzakas) [reply tests/lib/cli.js]: Good suggestion, I went with the second approach.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

File: tests/lib/cli.js [hunk 3]
```diff
@@ -237,10 +247,11 @@ describe("cli", () => {
 
         describe("Formatters", () => {
 
+            const flag = useFlatConfig ? "--no-config-lookup" : "--no-eslintrc";
+
             describe("when given a valid built-in formatter name", () => {
                 it(`should execute without any errors with configType:${configType}`, async () => {
                     const filePath = getFixturePath("passing.js");
-                    const flag = useFlatConfig ? "--no-config-lookup" : "--no-eslintrc";
                     const exit = await cli.execute(`${flag} -f checkstyle ${filePath}`, null, useFlatConfig);
```
Comment (mdjermanovic) [inline tests/lib/cli.js:255]:
```suggestion
                    const exit = await cli.execute(`${flag} -f json ${filePath}`, null, useFlatConfig);
```

**Review by mdjermanovic — COMMENTED** [review]
This might fix merge conflicts.

**Comment by nzakas** [comment]
Only CLA issues are with commits from CI, so this should be safe to merge.

**Comment by nzakas** [comment]
Will need to revisit once #17889 is merged

**Pull request merged**